Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp

```diff
@@ -386,7 +386,14 @@ void WebSWServerConnection::postMessageToServiceWorkerClient(ScriptExecutionCont
     if (!sourceServiceWorker)
         return;
 
-    send(Messages::WebSWClientConnection::PostMessageToServiceWorkerClient { destinationContextIdentifier, message, sourceServiceWorker->data(), sourceOrigin });
+    auto sourceServiceWorkerData = sourceServiceWorker->data();
+
+    sendWithAsyncReply(Messages::WebSWClientConnection::PostMessageToServiceWorkerClient { destinationContextIdentifier, message, sourceServiceWorkerData, sourceOrigin }, [this, weakThis = WeakPtr { *this }, destinationContextIdentifier, message, sourceServiceWorkerData, sourceOrigin] (bool result) {
+        if (result)
+            return;
+
+        server().addServiceWorkerClientPendingMessage(destinationContextIdentifier, { message, sourceServiceWorkerData, sourceOrigin });
```

Comment:
Looking at service worker container code, I see this patch is preserving the ordering so that is good.
One potential risk though is if ServiceWorkerContainer is not created (or not created immediately while parsing the document).
In that case, the messages would remain in the network process until ServiceWorkerContainer is created.
It seems that, once the document is parsed, if ServiceWorkerContainer is not created, we should just drop the messages.
This might be observable behaviour.

Reply:
The potential mis-ordering you suggest is accounted for:
Because the network process messages are known to be stored before web process messages, and therefore known to be fired before them.

(This occurs inside ServiceWorkerContainer::startMessages())